engine/execution/ingestion/stop_control.go

```diff
@@ -0,0 +1,221 @@
+package ingestion
+
+import (
+	"context"
+	"fmt"
+	"sync"
+
+	"github.com/rs/zerolog"
+
+	"github.com/onflow/flow-go/engine/execution/state"
+	"github.com/onflow/flow-go/model/flow"
+)
+
+// StopControl is a specialized component used by ingestion.Engine to encapsulate
+// control of pausing/stopping blocks execution.
+// It is intended to work tightly with the Engine, not as a general mechanism or interface.
+// StopControl follows states described in StopState
+type StopControl struct {
+	sync.RWMutex
+	height             uint64
+	crash              bool
+	stopAfterExecuting flow.Identifier
+	log                zerolog.Logger
+	state              StopControlState
+	lastExecutedHeight uint64 // last executed height, conveniently defaults to zero
+}
+
+type StopControlState byte
+
+const (
+	// StopControlOff default state, envisioned to be used most of the time. Stopping module is simply off,
+	// blocks will be processed "as usual".
+	StopControlOff StopControlState = iota
+
+	// StopControlSet means stop height is set but not reached yet, and nothing related to stopping happened yet.
+	// We can still go back to StopControlOff or progress to StopControlCommenced.
+	StopControlSet
+
+	// StopControlCommenced indicates that stopping process has commenced and no parameters can be changed anymore.
+	// For example, blocks at or above stop height has been received, but finalization didn't reach stop height yet.
+	// It can only progress to StopControlPaused
+	StopControlCommenced
+
+	// StopControlPaused means EN has stopped processing blocks. It can happen by reaching set stop height, or
+	// if the node was started with a pause mode.
+	// It is a final state and cannot be changed
+	StopControlPaused
+)
+
+// NewStopControl creates new empty NewStopControl
+func NewStopControl(log zerolog.Logger, paused bool) *StopControl {
```

Comment:
We need to initialize the highestProcessable with the HighestExecutedBlockHeight from badger database. So that we don't allow setting stop height lower or equal to that. 

Please refer to my comments before
```
// on startup, this is initialized as highest executed height
// after startup, this is updated by a Processable block with higher height 
// stop height can not be set below or equal to this height, otherwise we can't guarantee 
// no block above stopHeight has been executed. 
highestProcessableHeight uint64
```

Reply:
Added the initialization.